sklearn/cluster/_bisect_k_means.py

```diff
@@ -0,0 +1,727 @@
+
+        # Dictionary to imitate tree view of created centers.
+        # Used to keep hierarchical ordering.
+        tree_dict = {-1: {"children": None, "center": None}}
+
+        # Leaves Dictionary used for clustering.
+        # Stores information which data points are assigned to given cluster
+        # along with error or amount of assigned points, depending on
+        # specified 'bisect_strategy'
+        leaves_dict = {
+            -1: {"samples": np.ones(X.shape[0], dtype=bool), "error_or_size": None}
+        }
+
+        # ID of biggest center stored in centers_dict
+        biggest_id = -1
+
+        last_center_id = 0
+
+        for n_iter in range(self.n_clusters - 1):
+            picked_samples = leaves_dict[biggest_id]["samples"]
+
+            # Pick data and weights to bisect
+            picked_data = X[picked_samples]
+            picked_weights = sample_weight[picked_samples]
+
+            # Perform Bisection
+            _centers, _labels = self._bisect(picked_data, picked_weights, random_state)
+
+            if self.verbose:
+                print(f"Centroid Found: {_centers[0]}")
+                print(f"Centroid Found: {_centers[1]}")
+
+            if strategy == "biggest_sse":
+                # "biggest_sse" uses computed SSE (Sum of Squared Errors)
+                # to pick cluster with biggest SSE Error
+                metrics_values = self._compute_bisect_errors(
+                    picked_data, _centers, _labels, picked_weights
+                )
+            else:
+                # strategy == 'largest_cluster'
+                # 'largest_cluster' takes counts occurances of each label
+                # to pick cluster with largest number of data points assigned
+                metrics_values = np.bincount(_labels)
+
+            # "Create Hierarchy":
+            # Cluster with smaller metrics value (SSE or ammount of points)
+            # will be at 'left side' and cluster with higher at 'right side'
+            lower_id = 0 if metrics_values[0] <= metrics_values[1] else 1
+            centers_id = [lower_id, 1 - lower_id]
+
+            # Assign calculated nested clusters to their root
+            tree_dict[biggest_id]["children"] = [last_center_id, last_center_id + 1]
+
+            for id in range(2):
+                child_id = tree_dict[biggest_id]["children"][id]
+                picked_id = centers_id[id]
+
+                # Save Results on Tree
+                tree_dict[child_id] = {"children": None, "center": _centers[picked_id]}
+
+                # Create Mask for samples for selecting proper data points
+                samples_mask = leaves_dict[biggest_id]["samples"].copy()
+                samples_mask[picked_samples] = _labels == picked_id
+
+                # Save recently generated leaves
+                leaves_dict[child_id] = {
+                    "samples": samples_mask,
+                    "error_or_size": metrics_values[picked_id],
+                }
```

Comment:
I find the variables `lower_id`, `centers_id`, `id`, `child_id`, `biggest_id`, `last_center_id`, `picked_id` to be a bit confusing. To try to clarify them, I propose to keep "id" for the global cluster IDs, and to rename into "label" when it corresponds to the labels of the two clusters.

```suggestion
            ordered_labels = (0, 1) if metrics_values[0] <= metrics_values[1] else (1, 0)

            # Assign calculated nested clusters to their root
            tree_dict[biggest_id]["children"] = [last_center_id, last_center_id + 1]

            for ii, label in enumerate(ordered_labels):
                child_id = tree_dict[biggest_id]["children"][ii]

                # Save Results on Tree
                tree_dict[child_id] = {"children": None, "center": _centers[label]}

                # Create Mask for samples for selecting proper data points
                samples_mask = leaves_dict[biggest_id]["samples"].copy()
                samples_mask[picked_samples] = _labels == label

                # Save recently generated leaves
                leaves_dict[child_id] = {
                    "samples": samples_mask,
                    "error_or_size": metrics_values[label],
                }
```